File: PA1_Server/src/edu/ucdenver/tournament/Match.java
package edu.ucdenver.tournament;

import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.ArrayList;

/**
 * <p>This defines the Match class.
 * The Match class is responsible
 * for creating matches, setting lineups
 * setting referees, and match scores
 * </p>
 * @author David Desrochers
 * @version 1.0
 * @since 2022-11-11
 */
public class Match implements Serializable {

    /**
     * <p>
     *     Starting date time for the match
     * </p>
     */
    private final LocalDateTime dateTime;

    /**
     * <p>
     *     Team A Score
     * </p>
     */
    private int scoreTeamA;

    /**
     * <p>
     *     Team B Score
     * </p>
     */
    private int scoreTeamB;

    /**
     * <p>
     *     Line up A in the match
     * </p>
     */
    private final LineUp lineUpA;

    /**
     * <p>
     *     Line up B in the match
     * </p>
     */
    private final LineUp lineUpB;

    /**
     * <p>
     *     List of referees for the match.
     *     Must be a minimum of 4 for the match
     *     to begin
     * </p>
     */
    private final ArrayList<Referee> refereeList;

    /**
     * <p>
     *     If the match score has
     *     been recorded, set to true.
     * </p>
     */
    private boolean isRecorded;


    /**
     * <p>
     * This is the constructor
     * for the Match class. It sets
     * the date/time, and lineups for teams
     * A and B. This initializes the referee list.
     * </p>
     * @param dateTime  date/time of match
     * @param TeamA     Team A
     * @param TeamB     Team B
     */
    public Match(LocalDateTime dateTime, Team TeamA, Team TeamB) {
        this.dateTime = dateTime;
        this.lineUpA = new LineUp(TeamA);
        this.lineUpB = new LineUp(TeamB);
        this.refereeList = new ArrayList<>();
        this.isRecorded = false;
    }

    /**
     * <p>
     * Returns Team A
     * </p>
     * @return Team A
     */
    public LineUp getTeamA(){
        return this.lineUpA;
    }

    /**
     * <p>
     * Returns Team B
     * </p>
     * @return Team B
     */
    public LineUp getTeamB(){
        return this.lineUpB;
    }

    /**
     * <p>
     * Returns if a match is upcoming
     * </p>
     * @return True or False
     */
    public boolean isUpcoming(){
        return this.dateTime.isAfter(LocalDateTime.now());
    }

    /**
     * <p>
     * Adds a player to a lineup of a team.
     * Ensures player isn't already in lineup.
     * </p>
     * @param t Team Name
     * @param p Player Name
     */
    public void addPlayer(Player p, Team t){
        boolean flag = true;
        if (this.lineUpA.getTeam().equals(t)){
            for (Player player: this.lineUpA.getPlayers()){
                if (player.getName().equals(p.getName())) {
                    flag = false;
                    break;
                }
            }
            if (flag){
                this.lineUpA.addPlayer(p);
            }
        }
        else if (this.lineUpB.getTeam().equals(t)){
            for (Player player: this.lineUpB.getPlayers()){
                if (player.getName().equals(p.getName())) {
                    flag = false;
                    break;
                }
            }
            if (flag){
                this.lineUpB.addPlayer(p);
            }
        }
    }

    /**
     * <p>
     * Adds a referee to match.
     * Ensures referee count isn't above 4.
     * </p>
     * @param referee Referee object
     */
    public void addReferee(Referee referee) {
        this.refereeList.add(referee);
    }

    /**<p>
     * Sets the score for a match
     * </p>
     * @param A     Score of Team A
     * @param B     Score of Team B
     */
    public void setMatchScore(int A, int B){
        this.isRecorded = true;
        this.scoreTeamA = A;
        this.scoreTeamB = B;
    }

    /**
     * <p>
     * Returns Datetime of match
     * </p>
     * @return datetime
     */
    public LocalDateTime getDateTime() {
        return this.dateTime;
    }

    /**
     * <p>
     * Returns referees in match
     * </p>
     * @return referee list
     */
    public ArrayList<Referee> getRefereeList() {
        return this.refereeList;
    }

    /**
     * <p>
     * Returns Lineup A
     * </p>
     * @return Lineup A
     */
    public Country getLineUpACountry() {
        return this.lineUpA.getTeam().getCountry();
    }

    /**
     * <p>
     * Returns Lineup B
     * </p>
     * @return Lineup B
     */
    public Country getLineUpBCountry() {
        return this.lineUpB.getTeam().getCountry();
    }

    /**
     * <p>
     * Returns if a match is recorded
     * </p>
     * @return True or false
     */
    public boolean getIsRecorded(){
        return this.isRecorded;
    }

    /**
     * <p>
     * Returns if a match description listing
     * the team names and the time of the match.
     * Example:
     *      TeamA vs. TeamB on 2022-11-11 at 22:00
     * </p>
     * @return String of match description
     */
    @Override
    public String toString() {
        return lineUpA.getTeam().getName() + " vs. " + lineUpB.getTeam().getName() +
                    " on " + this.dateTime.toLocalDate() + " at " + this.dateTime.toLocalTime().toString();
    }

    /**
     * <p>
     * If the match is in the past, the score is included:
     * Example:
     *      TeamA vs. TeamB on 2022-11-11 at 22:00 --> Final Score: 2-1
     * </p>
     * @return String of match description
     */
    public String recordedToString(){
        return lineUpA.getTeam().getName() + " vs. " + lineUpB.getTeam().getName() +
                    " on " + this.dateTime.toLocalDate() + " at " + this.dateTime.toLocalTime() +
                    " --> Final Score: " + this.scoreTeamA + " - " + this.scoreTeamB;
    }
}

File: Server (Python)/src/Match.py
from datetime import datetime
import Country
from LineUp import LineUp
import Player
import Referee
import Team


class Match:
    """
    This defines the Match class.
    The Match class is responsible
    for creating matches, setting lineups
    setting referees, and match scores
    """
    _date_time: datetime  # start time of match
    _score_team_a: int  # team A score
    _score_team_b: int  # team B score
    _line_up_a: LineUp  # lineup A
    _line_up_b: LineUp  # lineup B
    _referee_list: list  # list of referees for match
    _is_recorded: bool  # is score recorded

    def __init__(self, _date_time: datetime, team_a: Team, team_b: Team):
        """
        This is the constructor
        for the Match class. It sets
        the date/time, and lineups for teams
        A and B. This initializes the referee list.
        :param _date_time: start time of match
        :param team_a: team A object
        :param team_b: team B object
        """
        self._date_time = _date_time
        self._line_up_a = LineUp(team_a)
        self._line_up_b = LineUp(team_b)
        self._referee_list = []
        self._is_recorded = False

    @property
    def get_team_a(self) -> LineUp:
        """
        Returns lineup A
        :return: lineup A (Lineup)
        """
        return self._line_up_a

    @property
    def get_team_b(self) -> LineUp:
        """
        Returns lineup B
        :return: lineup B (Lineup)
        """
        return self._line_up_b

    @property
    def get_start_time(self) -> datetime:
        """
        Returns start time of match
        :return: start time (datetime)
        """
        return self._date_time

    @property
    def get_referee_list(self) -> []:
        """
        Returns referee list
        :return: referee list (List)
        """
        return self._referee_list

    @property
    def get_lineup_a_country(self) -> Country:
        """
        Gets Lineup A country
        :return: Lineup A country (Country)
        """
        return self._line_up_a.get_team.get_country

    @property
    def get_lineup_b_country(self) -> Country:
        """
        Gets Lineup B country
        :return: Lineup B country (Country)
        """
        return self._line_up_b.get_team.get_country

    @property
    def get_is_recorded(self) -> bool:
        """
        Returns if the match is recorded
        :return: true or false (Bool)
        """
        return self._is_recorded

    def is_upcoming(self) -> bool:
        """
        Returns if the match is upcoming
        :return: true or false (Bool)
        """
        return self._date_time > datetime.now()

    def add_player(self, player: Player, team: Team):
        """
        Adds a player to a lineup
        :param player: player object
        :param team: team object
        """
        flag: bool = True

        if self._line_up_a.get_team == team:
            for p in self._line_up_a.get_players:
                if p.get_name == player.get_name:
                    flag = False
                    break
            if flag:
                self._line_up_a.add_player(player)

        elif self._line_up_b.get_team == team:
            for p in self._line_up_b.get_players:
                if p.get_name == player.get_name:
                    flag = False
                    break
            if flag:
                self._line_up_b.add_player(player)

    def add_referee(self, referee: Referee):
        """
        Adds a referee to a match
        :param referee: referee object
        """
        self._referee_list.append(referee)

    def set_match_score(self, a: int, b: int):
        """
        Sets the match score
        :param a: team A score
        :param b: team B score
        """
        if not self._is_recorded:
            self._is_recorded = True
            self._score_team_a = a
            self._score_team_b = b

    def __str__(self) -> str:
        """
        String representation of match
        :return: match string (String)
        """
        return self._line_up_a.get_team.get_name + " vs. " + self._line_up_b.get_team.get_name + " on " + \
               str(self._date_time.date()) + " at " + str(self._date_time.time())

    def recorded_to_string(self) -> str:
        """
        Recorded string representation of match
        :return: recorded match string (String)
        """
        return self._line_up_a.get_team.get_name + " vs. " + self._line_up_b.get_team.get_name + " on " + \
               str(self._date_time.date()) + " at " + str(self._date_time.time()) + " --> Final Score: " + \
               str(self._score_team_a) + " - " + str(self._score_team_b)
